Sources/SwiftSyntax/BumpPtrAllocator.swift

```diff
@@ -0,0 +1,114 @@
+//===---------- BumpPtrAllocator.swift - Bump-pointer Allocaiton ----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Bump-pointer allocation.
+struct BumpPtrAllocator {
+  typealias Slab = UnsafeMutableRawBufferPointer
+
+  static private var SLAB_SIZE: Int = 4096
+  static private var GLOWTH_DELAY: Int = 128
+  static private var SLAB_ALIGNMENT: Int = 8
+
+  private var slabs: [Slab]
+  private var currentPtr: UnsafeMutableRawPointer?
+  private var endPtr: UnsafeMutableRawPointer?
+  private var customSizeSlabs: [Slab]
+  private(set) var totalSizeAllocated: Int
+
+  init() {
+    slabs = []
+    currentPtr = nil
+    endPtr = nil
+    customSizeSlabs = []
+    totalSizeAllocated = 0
+  }
+
+  /// Calculate the size of the slab at the index.
+  private static func slabSize(at index: Int) -> Int {
+    // Double the slab size every 'GLOWTH_DELAY' slabs.
+    return SLAB_SIZE * (1 << min(30, index / GLOWTH_DELAY))
+  }
+
+  private mutating func startNewSlab() {
+    let newSlabSize = Self.slabSize(at: slabs.count)
+    let newSlab = Slab.allocate(
+      byteCount: newSlabSize, alignment: Self.SLAB_ALIGNMENT)
+    slabs.append(newSlab)
+    currentPtr = newSlab.baseAddress!
+    endPtr = currentPtr!.advanced(by: newSlabSize)
+  }
+
+  /// Allocate 'byteCount' of memory.
+  mutating func allocate(byteCount: Int, alignment: Int) -> UnsafeMutableRawBufferPointer {
+
+    precondition(alignment <= Self.SLAB_ALIGNMENT)
+    guard byteCount > 0 else {
+      return .init(start: nil, count: 0)
+    }
+
+    // Track the total size we allocated.
+    totalSizeAllocated += byteCount
+
+    // Check if the current slab have enough space.
+    if !slabs.isEmpty {
+      let aligned = currentPtr!.alignedUp(toMultipleOf: alignment)
+      if (aligned + byteCount <= endPtr!) {
+        currentPtr = aligned + byteCount
+        return .init(start: aligned, count: byteCount)
+      }
+    }
+
+    // If the size is too big, allocate a dedicated slab for it.
+    if (byteCount >= Self.SLAB_SIZE) {
```

Comment:
Same here

Reply:
This check needs to take into account that the requested size will differ from the allocation size.

Reply:
Although maybe it doesn't. The allocation will always be at least as large as the requested size given alignment constraints...

Reply:
Since I limited the max alignment to the slab alignment,
```
    precondition(alignment <= Self.SLAB_ALIGNMENT)
```
I thought we don't need to consider the alignment here. Am I missing anything?